03-Collections/Collections/Collections.cs

```diff
@@ -75,61 +100,128 @@ public class Task {
-        ///                     |
-        ///                     8   
-        ///                   
-        ///    result = { 1, 2, 3, 4, 5, 6, 7, 8 } 
-        /// </example>
-        public static IEnumerable<T> WidthTraversalTree<T>(ITreeNode<T> root) {
-            // TODO : Implement the tree width traversal algorithm
-            throw new NotImplementedException();
-        }
+			if (root == null)
+				throw new ArgumentNullException();
 
+			var deepFirst = new Stack<ITreeNode<T>>();
+			deepFirst.Push(root);
 
+			while (deepFirst.Count > 0)
+			{
+				var lastElement = deepFirst.Pop();
+				yield return lastElement.Data;
 
-        /// <summary>
-        ///   Generates all permutations of specified length from source array
-        /// </summary>
-        /// <typeparam name="T"></typeparam>
-        /// <param name="source">source array</param>
-        /// <param name="count">permutation length</param>
-        /// <returns>
-        ///    All permuations of specified length
-        /// </returns>
-        /// <exception cref="System.InvalidArgumentException">count is less then 0 or greater then the source length</exception>
-        /// <example>
-        ///   source = { 1,2,3,4 }, count=1 => {{1},{2},{3},{4}}
-        ///   source = { 1,2,3,4 }, count=2 => {{1,2},{1,3},{1,4},{2,3},{2,4},{3,4}}
-        ///   source = { 1,2,3,4 }, count=3 => {{1,2,3},{1,2,4},{1,3,4},{2,3,4}}
-        ///   source = { 1,2,3,4 }, count=4 => {{1,2,3,4}}
-        ///   source = { 1,2,3,4 }, count=5 => ArgumentOutOfRangeException
-        /// </example>
-        public static IEnumerable<T[]> GenerateAllPermutations<T>(T[] source, int count) {
-            // TODO : Implement GenerateAllPermutations method
-            throw new NotImplementedException();
-        }
+				if (lastElement.Children != null)
+				{
+					var arrayChildren = lastElement.Children.ToArray().Reverse();
+					foreach (var item in arrayChildren)
+					{
+						deepFirst.Push(item);
+					}
+				}
+			}
+		}
 
-    }
+		/// <summary>
+		///   Traverses a tree using the width-first strategy
+		/// </summary>
+		/// <typeparam name="T">tree node type</typeparam>
+		/// <param name="root">the tree root</param>
+		/// <returns>
+		///   Returns the sequence of all tree node data in width-first order
+		/// </returns>
+		/// <example>
+		///    source tree (root = 1):
+		///    
+		///                      1
+		///                    / | \
+		///                   2  3  4
+		///                  / \     \
+		///                 5   6     7
+		///                     |
+		///                     8   
+		///                   
+		///    result = { 1, 2, 3, 4, 5, 6, 7, 8 } 
+		/// </example>
+		public static IEnumerable<T> WidthTraversalTree<T>(ITreeNode<T> root) {
+			// TODO : Implement the tree width traversal algorithm
+
+			if (root == null)
+				throw new ArgumentNullException();
+
+			Queue<ITreeNode<T>> widthFirst = new Queue<ITreeNode<T>>();
```

Comment:
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions
see the part about Implicitly Typed Local Variables. Fix it for all such declarations in the file